cdap-storage-spi/src/main/java/io/cdap/cdap/spi/data/StructuredTableContext.java

```diff
@@ -36,4 +39,18 @@
    */
   StructuredTable getTable(StructuredTableId tableId)
     throws StructuredTableInstantiationException, TableNotFoundException;
+
+  /**
+   * Get an instance of the specified Dataset. If none exist, create one based on the given
+   * {@link StructuredTableSpecification} first and then returns it.
+   *
+   * @param tableId the table id
+   * @return An instance of the specified table, never null
+   * @throws StructuredTableInstantiationException if the table cannot be instantiated
+   * @throws TableNotFoundException if table was deleted after existence check, therefore leading to table not found
+   * @throws IOException if table creation failed
+   */
+  StructuredTable getOrCreateTable(StructuredTableId tableId, StructuredTableSpecification tableSpec)
```

Comment:
this is a pretty risky API to add, as there are many ways to misuse it. For example, what happens if the table already exists, but has a different spec than the one passed in? What about race conditions? What happens if somebody tries to add a huge number of tables? 

auto-creation of the tables seem like it should be something that happens behind the scenes based on how CDAP is setup. This method requires that the app knows something about the CDAP setup, and actually makes the existing getTable() useless.

Reply:
I'm not sure what the best way to implement this would be, but one possibility is to record the table specs in the ApplicationSpecification (if they aren't there already) and auto-create them on startup in UserServiceProgramMain if CDAP is configured to use leveldb.

Another possibility could be to make a table admin that makes remote calls to appfabric to get table specs, but local calls to create tables. Not sure how feasible this is.

Reply:
In either scenario, I'm not sure how updates to the table would be handled. Though we don't have that use case today, I would imagine it will come at some point in the future.

Reply:
This pattern is what we want to avoid when we refactor the storage. We want to have a central place to create all the tables. Previously we have this kind of pattern in storage and we create table when needed, that gives us a headache to find out what each store is doing, we should avoid introducing them again. 
The table spec is stored in the registry table I believe, if a spec does not match, we should fail the creation.